### DISCUSSION TITLE
Help with umqtt.simple MQTTClient init parameters version1.4:  how to create a SSL connection?

### DISCUSSION THREAD

**Discussion by Jcspoza** [micropython/micropython#13624] [RP2040 / Pico] [answered]
Hello
I am new with mqtt, so I follw the excellente tutorial:
[Getting Started with Raspberry Pi Pico W for IoT: Micropython and MQTT](https://www.hivemq.com/article/iot-reading-sensor-data-raspberry-pi-pico-w-micropython-mqtt-node-red/)

I have installed the library umqtt.simple using Thonny package management feature. If you install version 1.3 of umqtt.simple, everything was perfectly OK. But if you install version 1.4 of umqtt.simple, the way to manage SSL does not work, because tutorial ( I guess) is based on umqtt.simple version 1.3 with a this initialisation parameters schema like :
`def connectMQTT():
    client = MQTTClient(client_id=b"kudzai_raspberrypi_picow",
    server=b"8fbadaf843514ef286a2ae29e80b15a0.s1.eu.hivemq.cloud",
    port=0,
    user=b"mydemoclient",
    password=b"passowrd",
    keepalive=7200,
    ssl=True,
    ssl_params={'server_hostname':'8fbadaf843514ef286a2ae29e80b15a0.s1.eu.hivemq.cloud'}
    )
`
But in umqtt.simple version 1.4 , ssl_params has disappeared, and ssl has another type not a simple boolean. i.e 
`def connect(self, clean_session=True):
        self.sock = socket.socket()
        addr = socket.getaddrinfo(self.server, self.port)[0][-1]
        self.sock.connect(addr)
        if self.ssl:
            self.sock = self.ssl.wrap_socket(self.sock, server_hostname=self.server)
......
`
see 'ssl.wrap_socket(self.sock, server_hostname=self.server)'

In conclusion , I need help to create a mqtt client with SSL = true, and the rest of configurations using umqtt.simple version 1.4

Thanks in advance!!

**Comment by felixdoerre** [comment, marked as answer]
Yes, the way how tls is handled has changed. For umqtt.simple 1.4 you need to pass in the ssl context yourself:
```
context = ssl.SSLContext(ssl.PROTOCOL_TLS_CLIENT)
context.verify_mode = ssl.CERT_NONE
client = MQTTClient(client_id=b"kudzai_raspberrypi_picow", server=b"8fbadaf843514ef286a2ae29e80b15a0.s1.eu.hivemq.cloud", port=0, user=b"mydemoclient", password=b"passowrd", keepalive=7200, ssl=context )
```

**Reply by Jcspoza** [reply]
Thanks a lot!! It works!

Details 
- micropython version : MicroPython v1.23.0-preview.114.gb4f59984f on 2024-02-09; Raspberry Pi Pico W with RP2040
- umqtt.simple version : 1.4.0 installed with Thonny version 4.1.4 ( previosu version show problems with mqtt libs)
- code
```
# 3- Creacion objeto cliente MQTT & conexion a broker
context = ssl.SSLContext(ssl.PROTOCOL_TLS_CLIENT)
context.verify_mode = ssl.CERT_NONE
try:
    client = MQTTClient(client_id=b"jcsp_picow", # something unique to identify device, ie machine.unique_id()
                        server= MQTT_SERVER,
                        port=0, # if port ==0 AND ssl=True -> port=8883 the secure mqtt port
                        user=secrets['mqttUser'],
                        password=secrets['mqttPassword'],
                        keepalive=7200,
                        ssl=context) 

    client.connect()
    print('Connected to %s MQTT Broker'%(MQTT_SERVER))
except OSError as e:
    print('Failed to connect to the MQTT Broker.')
    time.sleep(5)
    machine.reset()
```

Unfortunately, I have a side-efect with the miocropython version I want to use wich is a firmware provided by Russ Hughes @russhughes to manage ST7735, ST7789 and ILI 9431 displays!! This firmware is based on a 1.20 micropython version from October 2023, so ssl lib is the previous one, I guess:  I tested with umqtt.simple 1.4 -> error 
```
context = ssl.SSLContext(ssl.PROTOCOL_TLS_CLIENT)
```
AttributeError: 'module' object has no attribute 'SSLContext'

So I will continue using umqtt.simple version 1.3 since the changes I must to do in display part of my project are greather than in mqtt connection part. In my next mqtt project I will try to employ new version.

Thanks
